### PR TITLE
fix dictionary keys changed during iteration error seen in utils when using py3.8

### PR DESCRIPTION
**Pull request by elinnore**
noticed this error when trying to upload a file in jira while using python 3.8 on MacOS 10.14.6:
```
Python 3.8.0 (default, Nov 25 2019, 19:38:49) 
[Clang 10.0.1 (clang-1001.0.46.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from jira import JIRA
>>> jira = JIRA(server='http://localhost:8080', basic_auth=('myuser', 'mypass'))
>>> attached = jira.add_attachment(issue='CSTREQ-1', attachment='file_with_text.log')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/myuser/.pyenv/versions/venv/lib/python3.8/site-packages/jira/client.py", line 126, in wrapper
    result = func(*arg_list, **kwargs)
  File "/Users/myuser/.pyenv/versions/venv/lib/python3.8/site-packages/jira/client.py", line 787, in add_attachment
    url, data=m, headers=CaseInsensitiveDict({'content-type': m.content_type, 'X-Atlassian-Token': 'nocheck'}), retry_data=file_stream)
  File "/Users/myuser/.pyenv/versions/venv/lib/python3.8/site-packages/jira/utils/__init__.py", line 41, in __init__
    for key, value in super(CaseInsensitiveDict, self).items():
RuntimeError: dictionary keys changed during iteration
```
Tested the fix below and there seems to not be any issues. File uploaded fine, no other errors:
```
>>> attached = jira.add_attachment(issue='CSTREQ-1', attachment='file_with_text.log')
>>> attached
<JIRA Attachment: filename='file_with_text.log', id='10000', mimeType='text/plain'>
```

### CODE REVIEW

File: jira/utils/__init__.py [hunk 1]
```diff
@@ -36,13 +36,18 @@ class CaseInsensitiveDict(dict):
     def __init__(self, *args, **kw):
         super(CaseInsensitiveDict, self).__init__(*args, **kw)
 
-        self.itemlist = {}
-        for key, value in super(CaseInsensitiveDict, self).items():
-            if key != key.lower():
-                self[key.lower()] = value
-                self.pop(key, None)
+#        self.itemlist = {}
```
Comment (lan17) [inline jira/utils/__init__.py]: Do not leave commented out code.  Remove it if it's no longer necessary.

File: jira/utils/__init__.py [hunk 2]
```diff
@@ -36,13 +36,18 @@ class CaseInsensitiveDict(dict):
...
+#        for key, value in super(CaseInsensitiveDict, self).items():
+#            if key != key.lower():
+#                self[key.lower()] = value
+#                self.pop(key, None)
 
         # self.itemlist[key.lower()] = value
+        itemlist = {}
+        for key, value in super(CaseInsensitiveDict, self).items():
+            itemlist[key.lower()] = value
```
Comment (lan17) [inline jira/utils/__init__.py]: Why not use a list of keys to remove and remove them in subsequent for loop?

Comment (elinnore) [reply jira/utils/__init__.py]: thank you. Yes, that would be clearer actually and appears to work fine too:
```
    def __init__(self, *args, **kw):
        super(CaseInsensitiveDict, self).__init__(*args, **kw)

        upper_keys_list = []
        for key in super(CaseInsensitiveDict, self).keys():
            if key != key.lower():
                upper_keys_list.append(key)

        for upper_key in upper_keys_list:
            self[upper_key.lower()] = self[upper_key]
            self.pop(upper_key, None)
```

**Comment by yhuskov** [comment]
Can this pull request be reopened? The bug prevents using the `add_attachment` method with Python 3.8.

**Comment by keyroii** [comment]
`add_attachment` method still not working in python 3.8 as of may.

**Comment by dbaxa** [comment]
cc @ssbarnea (this seems like a candidate for review & merge).

**Comment by ackris** [comment]
It is still not working. Please find my [comment](https://github.com/pycontribs/jira/issues/890#issuecomment-625201046) on [Issue ](https://github.com/pycontribs/jira/issues/890). Please do a review and release a working candidate

**Comment by olologin** [comment]
Oh come on, is it so hard to merge this? It still fails on latest master.

**Comment by dbaxa** [comment]
@ssbarnea Can we merge this :-) ?

**Comment by AnatoliiSliusarievSecurrency** [comment]
Can this be merged? It's really a blocker

**Comment by wesinator** [comment]
seems the issue was already fixed in https://github.com/pycontribs/jira/pull/895 , just hasn't been released

**Comment by dvaerum** [comment]
> seems the issue was already fixed in #895 , just hasn't been released

Can we please have a release when? 🥺 I mean it have been 1/2 year

**Comment by wesinator** [comment]
> > seems the issue was already fixed in #895 , just hasn't been released
> 
> Can we please have a release when? 🥺 I mean it have been 1/2 year

https://github.com/pycontribs/jira/releases/tag/3.0

**Comment by sdanelli** [comment]
@russoz I think we can help as we are also affected, do you have the details on what is broken to ship this fix?

**Comment by russoz** [comment]
@sdanelli I am not a maintainer of the project, but issue https://github.com/pycontribs/jira/issues/896 is pinned on the Issues tab - the project maintainers are asking for help.

**Comment by jd-ssg** [comment]
Could we merge this fix, please ?

**Comment by adehad** [comment]
No idea if people are still following this, but I've made a PR for the move to `requests.structures.CaseInsensitiveDict` #1084 which I think also addresses this, I've added a test in which suggests it does, but the test also passes before the change so  would be great to confirm with others.

**Pull request closed without merging**